New custom option - "Stay in current window if compiler errors are detected."

jhegedus42 opened the pull request.

hannesm commented on idris-commands.el:
```diff
@@ -238,7 +238,11 @@ line."
+                 (display-buffer (idris-buffer-name :notes))
+                 (pop-to-buffer (idris-buffer-name :notes))
+	       )
+	       )))))
```
thx - but please follow the emacs lisp syntax rules used in this repository, namely: no whitespace before `if`, no newline before closing parens...

jhegedus42 replied:
yes, will do - thanks for pointing it out

jhegedus42 replied:
How do I correct that ? Do I just push again and create a new PR ?

ahmadsalim replied:
@jhegedus42 You would normally just push again to your branch.

hannesm replied:
please force-push to your branch

jhegedus42 replied:
Pushed. Hope this is ok.

hannesm commented on idris-commands.el:
```diff
@@ -238,7 +238,10 @@ line."
+	       (if idris-stay-in-current-window-on-compiler-error
+                 (display-buffer (idris-buffer-name :notes))
+                 (pop-to-buffer (idris-buffer-name :notes)))
+	       )))))
```
there's still a newline before the `)))))`... and can you please squash the commits into a single one? thanks!

jhegedus42 replied:
I am quite new to git, how do I squash the commits into a single commit? I guess I need to create a new PR for that right?

hannesm replied:
no need to create a new PR, do your commits and then a git rebase -i origin/master, mark all but the first one as fixup.  and then force-push

david-christiansen commented:
I also think it might be good to make this setting a bit more general. For example, I think it should cover the function used to display the hole list, and perhaps all of the info displays.

jhegedus42 commented:
yeah, the hole list is also an often used one, which function is that ?

david-christiansen commented:
Looks like this is already using `display-buffer`. I think that's fine.

https://github.com/idris-hackers/idris-mode/blob/master/idris-hole-list.el#L103

What about the other uses of `pop-to-buffer`? Which of them should, in your opinion, be customizable?

Finally, please don't use cryptic commit messages like "Cosmetix", which sounds to me like an Asterix and Obelix reference rather than a description of what the commit does. It would be better to write something like "Match project code style conventions".

[Here's an intro](https://davidwalsh.name/squash-commits-git) to squashing commits.

jhegedus42 commented:
Oh, ok, right, I am quite new to this github things, I see your point. Yeah, what other uses of pop-to-buffer should be customizable ? Well atm this was the really big thing for me, coz I make lot of mistakes and I needed to re-focus after type checking about 100 times per day :) - nothing else comes to mind (the other pop-to-buffer uses don't cause any issue because I rarely use those features).

david-christiansen commented:
It's perfectly fine to be new to GitHub and working on projects - we'll do our best to help you get up to speed!

How do you feel about the documentation buffer being part of this customization? That is, if you do `C-c C-d d` on a name and get its docs, would you want that to just be show in another window without the focus going there? This is how Emacs's built-in docs work, so it seems like it's related.

Oh, and one more thing - would you mind updating CHANGES.markdown in the root of the repository with a little note describing the new settings variable?

Thanks for being a good sport and for persisting while we ask for changes!

jhegedus42 commented:
The doc buffer seems to be just fine. I mean, it jumps to the doc window the first time but after that subsequent uses of `C-c C-d d` keep the Idris code focused (i.e. the window in which I hit the key-combo).

One other thing that might be handy is some kind of "go-to-definition". 


Sure, I update the CHANGES file too.

jhegedus42 commented:
I dunno, this squashing maybe somehow did not work... or did it ? How can I see if it worked ? I am not sure.

jhegedus42 commented:
```
You must specify --onto when using --root
>vi CHANGES.markdown ^C
>git rebase -i
Cannot rebase: You have unstaged changes.
Please commit or stash them.
>git rebase -i ^C
>git commit -a
[master 460741f] Code beutify.
 1 file changed, 1 insertion(+), 2 deletions(-)
>git rebase -i
fatal: Not a valid object name 460740f
fatal: Needed a single revision
Invalid commit name: 460740f
>git rebase -i

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
        git rebase (--continue | --abort | --skip)
If that is not the case, please
        rm -fr /home/joco/.emacs.d/elpa/idris-mode-20151030.407/.git/rebase-merge
and run me again.  I am stopping in case you still have something
valuable there.
>git commit -a
Aborting commit due to empty commit message.
>git checkout master
Previous HEAD position was f2578cb... Cosmetix.
Switched to branch 'master'
Your branch is ahead of 'origin/master' by 2 commits.
>git commit -a
# On branch master
# Your branch is ahead of 'origin/master' by 2 commits.
#
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
#
#       idris-mode-autoloads.el
#       idris-mode-pkg.el
nothing added to commit but untracked files present (use "git add" to track)
>git push
Counting objects: 9, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (6/6), done.
Writing objects: 100% (6/6), 876 bytes, done.
Total 6 (delta 4), reused 0 (delta 0)
To git@github.com:jhegedus42/idris-mode.git
   f2578cb..460741f  master -> master
>git rebase -i origin

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
        git rebase (--continue | --abort | --skip)
If that is not the case, please
        rm -fr /home/joco/.emacs.d/elpa/idris-mode-20151030.407/.git/rebase-merge
and run me again.  I am stopping in case you still have something
valuable there.
>git rebase -i master

It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
        git rebase (--continue | --abort | --skip)
If that is not the case, please
        rm -fr /home/joco/.emacs.d/elpa/idris-mode-20151030.407/.git/rebase-merge
and run me again.  I am stopping in case you still have something
valuable there.
>git rebase --continue
cat: /home/joco/.emacs.d/elpa/idris-mode-20151030.407/.git/rebase-merge/stopped-sha: No such file or directory
Successfully rebased and updated refs/heads/master.
>git push -f
Everything up-to-date
>
```

this is what I got ... no idea if this is OK or not.

hannesm commented:
look into `git status`, it should say something along the lines:
```
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
```

in case it does not, maybe it has a hint to run `git rebase --abort` (or continue, but let us abort).

once it is in this state, and `git diff` doesn't show any modifications, do a `git rebase -i origin/master` and an interactive buffer will appear.  Edit all commits apart from the first one to say `fixup` (instead of `commit`) -- thus all will be squashed into a single commit.

afterwards do a `git push -f`

jhegedus42 commented:
This is what I get with `git rebase -i origin/master` in VIM:

```
  1 noop
  2
  3 # Rebase 460741f..460741f onto 460741f
  4 #
  5 # Commands:
  6 #  p, pick = use commit
  7 #  r, reword = use commit, but edit the commit message
  8 #  e, edit = use commit, but stop for amending
  9 #  s, squash = use commit, but meld into previous commit
 10 #  f, fixup = like "squash", but discard this commit's log message
 11 #  x, exec = run command (the rest of the line) using shell
 12 #
 13 # If you remove a line here THAT COMMIT WILL BE LOST.
 14 # However, if you remove everything, the rebase will be aborted.
 15 #
```

hmmm.... I cannot really see any commits here.

jhegedus42 commented:
`git status` says:

```
>git status
# On branch master
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
#
#       idris-mode-autoloads.el
#       idris-mode-pkg.el
nothing added to commit but untracked files present (use "git add" to track)
>git diff
>git rebase -i origin/master
Successfully rebased and updated refs/heads/master.
>
```

jhegedus42 commented:
I can create a new PR if that helps. Lemme know.

hannesm commented:
sorry, my fault... I usually have the real upstream as `origin`, while you have your clone as `origin`.  This is fine.

To get to the solution - a single commit with all of your changes, the easiest way is to `git reset --soft HEAD~3` (which will undo the last 3 commits, but leave the modified files around) and a `git commit --amend .` to add the modifications to the first commit.

Afterwards, `git push -f`.

thanks

jhegedus42 commented:
It worked :) thanks for the patience.

hannesm commented on idris-commands.el:
```diff
@@ -238,7 +238,9 @@ line."
              (when (member 'warnings-tree idris-warnings-printing)
                (idris-list-compiler-notes)
-               (pop-to-buffer (idris-buffer-name :notes)))))))
+	       (if idris-stay-in-current-window-on-compiler-error
```
here's now a tab character introduced - which we also prefer to not have in this repository... sorry for being picky again

jhegedus42 replied:
tab? you mean at line 241 ?

hannesm replied:
yes. the indentation here on github doesn't look right, and when I try to mark it with a mouse, it seems to be a tab character

jhegedus42 replied:
ok, i will have a look after i slept a little ...

jhegedus42 replied:
i think i managed to get rid of the tab

hannesm commented:
imho this is fine (apart from the tab), but I'll leave it to @david-christiansen to decide whether he prefers some other `pop-to-buffer` being guarded (imho can be done in subsequent PRs as well)

david-christiansen commented:
Yeah, future ones can be a different setting as well.  Looks good!

david-christiansen commented:
Thanks, @jhegedus42!
